Lint: Update headers and checks per current guidance & provide helpful feedback

CAM-Gerlach opened the pull request:
Over the past >year, we've made significant progress in programmatically parsing the PEP headers, using them to intelligently display more useful, informative and readable output in the rendered PEPs, conforming them to our modern guidance, automatically checking their format and (particularly now with #2475) making it easy for other tools to consume and enrich them.

As a final step toward these overarching goals, this PR:

* Adds the previously-missing automatic checks for all the remaining headers, so errors in them that could affect rendering and human- and machine-readability no longer pass silently
* Updates the existing checks to reflect the current format specified by PEP 1/12 and expected by current and future tooling
* Conforms the modest number of remaining existing PEPs to ensure the headers render correctly and are machine-readable, and
* Provides much more helpful, specific and user-friendly descriptive check text.

Overall, this PRs enhancements to our existing checks:

* Helps PEP authors, by giving them near-instant, automatic and targeted feedback about any syntax issues, locally or on CI, without having to for a full rebuild or a human review
* Helps PEP editors by freeing them from the need to manually inspect and fix PEP headers
* Avoids any edge cases in our rendering pipeline
* Helps both readers and tools by ensuring the output is easy for both humans and machines to read, and
* Opens the door to more "smart" processing in the future (e.g. automatic Discussions-To generation) to reduce PEP author/editor workload while making the displayed output more useful and readable.

To note, this PR does **not** make any new headers required, invalidate any existing established header formats nor require any of the newer ones; the new checks only trigger on the formats.

Also, resolves the linting side of #2402 and supports the improvements in #2475 and #2467

hugovk commented on pep-0628.txt:
```diff
@@ -9,7 +9,6 @@ Content-Type: text/x-rst
 Created: 28-Jun-2011
 Python-Version: 3.6
 Post-History: 28-Jun-2011
-Resolution: https://bugs.python.org/issue12345
```
Looks like this thread does include the resolution, Guido's "I'm just going to do this" at https://bugs.python.org/issue12345#msg272287 ?

CAM-Gerlach replied:
I see; I'd elided the link from here since it didn't link directly to any resolution message, just a BPO issue, was the one remaining special-case for both the rendering and validation logic for this field, and was already mentioned and linked in the `PEP Acceptance` notice just below it.

In any case, thanks to your feedback, I went ahead and linked `accepted` in that block directly to the message above, while retaining the existing link to the issue itself.

JelleZijlstra replied:
Worth noting that BPO will move to GitHub soon (expected next weekend). There will be redirects provided, but not sure they'll work for individual messages.

CAM-Gerlach replied:
My understanding from following that is that redirects will be available, but they won't be the current URLs, and BPO won't be shut down, it will only become read-only. So this URL _should_ continue to work, like the many others in PEPs. If per-message linking stopped working, that would be a pretty significant breakage. See [@hugovk 's reply](https://discuss.python.org/t/github-issues-migration-is-coming-soon/13791/30?u=cam-gerlach) to my question in the thread, quoting psf/gh-migration#17 ,

However, I understand its not 100% clear from either the [Discourse thread](https://discuss.python.org/t/github-issues-migration-is-coming-soon/13791/28) or the devguide documentation (which took quite a bit of digging in the PR preview to even find). I can ask @ambv to confirm there.

gvanrossum replied:
Maybe while we wait for GitHub to allow the migration we could update the PEP for the migration with details like which urls will redirect where?

ezio-melotti replied:
If you want to link to the message directly you basically have two options:
* keep the current bpo link, that will continue to work for the foreseeable future (bpo will be around in read-only mode);
* wait until the migration is complete, and replace the bpo link with a link to the corresponding GitHub comment.

CAM-Gerlach replied:
> keep the current bpo link, that will continue to work for the foreseeable future (bpo will be around in read-only mode)

Given PEPs are historical documents, I don't see a problem with retaining the BPO links for existing, non-Draft/Active PEPs.

On a related note, I was thinking of adding `:bpo:` and `:gh;` roles, based on the current `:issue:` and `:gh:` in the CPython documentation, much like we do for `:pep:` and `:rfc:`. This would make linking to BPO and GitHub issues and PRs easier (just enter the number and it does the rest), while allowing us to make the render nicely and consistently, and most importantly easily update where the links point and how they render without having to touch any individual PEP or require authors to do anything extra. For example, we could (if we wanted/needed) switch `:bpo:` to point to the BPO redirect service to automatically redirect to the corresponding GitHub issue.

hugovk commented on pep-3145.txt:
```diff
@@ -2,7 +2,7 @@ PEP: 3145
 Title: Asynchronous I/O For subprocess.Popen
 Version: $Revision$
 Last-Modified: $Date$
-Author: (James) Eric Pruitt, Charles R. McCreary, Josiah Carlson
```
I don't think we should change people's names without consultation.

CAM-Gerlach replied:
I didn't want to change this, but unfortunately as `()` is the key element of the legacy authors email syntax, this breaks parsing and validation (at least without a bunch of hacks and/or special cases in the checkers and every tool that parses it). I meant to ping @ericpruitt here, but unfortunately it slipped my mind due to how late (≈2 am local) I finally finished this.

@ericpruitt , how do you want your name to be written here, preferably without parenthesis if that's okay? Just Eric Pruitt, as you list on your GitHub profile (as it currently is updated to)? Or J. Eric Pruitt, James Eric Pruitt or something else?

ericpruitt replied:
Please write it as just "Eric Pruitt".

CAM-Gerlach replied:
Thanks, that's what it's updated it to now :+1:

hugovk reviewed (commented):
Thanks for all this! A few comments.

JelleZijlstra commented on .pre-commit-config.yaml:
```diff
@@ -35,8 +33,8 @@ repos:
 
-      - id: check-ast
-        name: "Check Python AST"
+      # - id: check-ast
```
Should be removed or uncommented

CAM-Gerlach replied:
I commented it as a stopgap fix for #2402 , but depending on what we ultimately decide there, there's a chance we may want to restore it. That said, I can remove it, if you think its important.

JelleZijlstra commented on pep-0225.txt:
```diff
@@ -4,13 +4,13 @@ Version: $Revision$
 Python-Version: 2.1
 Post-History:
-Resolution: :pep:`PEP 465 -- Rejected alternatives to adding a new operator <465#rejected-alternatives-to-adding-a-new-operator>`
+Superseded-By: 465
```
I assume you mean 211, 225

CAM-Gerlach replied:
I ended up just reverting these changes since its not really the intended use of the `Superseded` status per PEP 1, and instead moved this to an explanatory note at the top of the PEP. I also added a missing `Resolution` link to the aforementioned PEP 465 to the message officially accepting it.

JelleZijlstra commented on pep-0331.txt:
```diff
@@ -2,7 +2,7 @@ PEP: 331
 Title: Locale-Independent Float/String Conversions
 Version: $Revision$
 Last-Modified: $Date$
-Author: Christian R. Reis <kiko at async.com.br>
```
Yes, can we just allow this in the regex?

CAM-Gerlach replied:
Actually, all email addresses are already automatically obfuscated to this format, as they always have been. Some authors might have thought they needed to do so manually from seeing the rendered PEPs or outdated guidance, but there's no need for the extra work for authors, complexity for parsing/linting and inconsistency in the source. (Sure, theoretically scrapers could dig through the entire GitHub source tree to find it, but if they're that sophisticated and comprehensive, they can likely also detect the common and fairly unambiguous `<email at domain.tld>` pattern anyway using a only slightly modified regex, just as we're doing ourselves here.)

Its not that hard to add to the linters, it just requires some time to test the additional regex against the suite of test cases and rewrite and/or revert parts of four different comments, when its automatically anonymized anyway. But if you really think its necessary, I certainly can.

JelleZijlstra commented on pep-0387.txt:
```diff
@@ -3,8 +3,8 @@ Title: Backwards Compatibility Policy
 Version: $Revision$
 Last-Modified: $Date$
 Author: Benjamin Peterson <benjamin@python.org>
-BDFL-Delegate: Brett Cannon (on behalf of the steering council)
```
Since this PEP is still "Active" it seems reasonable to change it

CAM-Gerlach replied:
Yeah, I didn't mass-change it; in this case though this PEP is explicitly stated to be approved `on behalf of the Steering Council` (rather than the BDFL) so I wanted to ensure that the header reflected that.

JelleZijlstra reviewed (commented):
Thanks for all the work that went into this!

@hugovk already had some good feedback and I have a few more comments. If possible, I'd like to avoid adding email addresses for people who previously didn't have one, since they may have had reasons not to want their email address here.

CAM-Gerlach commented on pep-0255.txt:
```diff
@@ -5,7 +5,6 @@ Last-Modified: $Date$
 Author: nas@arctrix.com (Neil Schemenauer),
         tim.peters@gmail.com (Tim Peters),
         magnus@hetland.org (Magnus Lie Hetland)
-Discussions-To: python-iterators@lists.sourceforge.net
```
Yep, I'd seen that, and they also mentioned it should be removed from the PEP there, so I went ahead and did so for that among other reasons.

CAM-Gerlach commented on pep-0245.txt:
```diff
@@ -3,7 +3,6 @@ Title: Python Interface Syntax
 Version: $Revision$
 Last-Modified: $Date$
 Author: Michel Pelletier <michel@users.sourceforge.net>
-Discussions-To: http://www.zope.org/Wikis/Interfaces
```
I'd poked around a bit and could find one that actually resolved to the apparently intended page, and if did, it would mean I'd have to add a special case for this, since every other `Discusses-To` link/email both validates via the regex and parses and renders properly under the current logic in #2467 .

taleinat replied:
The latest version of this on the Wayback Machine is [this from April 2006](https://web.archive.org/web/20060410173246/http://zope.org/Wikis/Interfaces/FrontPage) (that is after a redirect from [here](https://web.archive.org/web/20050415141542/http://www.zope.org/Wikis/Interfaces)).

taleinat replied:
In general though, IMO it's worth keeping the original URI even if it's no longer available.

CAM-Gerlach replied:
Great, thanks for the find. Since this is a wiki page with resources rather than a discussion venue, would require special casing both the header parsing and validation logic, and we can provide some additional context thanks to your efforts, I moved it to a brief historical note at the top of the PEP, that includes the link to the archive as well as the original, and also links the archive of the now also-gone `interface-dev` mailing list linked there that contains much of the discussion leading to the PEP.

CAM-Gerlach commented on pep-0211.txt:
```diff
@@ -3,13 +3,13 @@ Title: Adding A New Outer Product Operator
 Python-Version: 2.1
 Post-History:
-Resolution: :pep:`PEP 465 -- Rejected alternatives to adding a new operator <465#rejected-alternatives-to-adding-a-new-operator>`
+Superseded-By: 465
```
See above

CAM-Gerlach commented on .pre-commit-config.yaml:
```diff
@@ -77,59 +75,108 @@ repos:
+        files: '^pep-\d+\.(rst|txt)$'
+        exclude: '^pep-(0499)\.(rst|txt)$'
+        types: [text]
+      - id: validate-author
```
Done, thanks.

CAM-Gerlach commented on pep-0548.rst:
```diff
@@ -2,7 +2,7 @@ PEP: 548
 Title: More Flexible Loop Control
 Version: $Revision$
 Last-Modified: $Date$
-Author: R David Murray
```
After initial testing with it required, I ensured the email address was optional in all header checks (previously, I special-cased the few instances where authors had omitted it—only a couple in the modern `.rst` era).

I only added it to authors that had already listed it in their other PEPs, and thus seemed to be a simple oversight. However, looking back through my hits, I noticed that unlike the others, this email was linked in another PEP (and is also, incidentally, publicly listed on their GitHub profile), but not actually listed in the author, so I rebased it out, and ensured this wasn't the case for any others.

CAM-Gerlach commented:
I don't understand what's going on with GH Actions here. The check suite succeeds locally, and the problem that's causing the failure (introduced in #2489) is not on this branch (which I've checked and double-checked). Did GH Actions change to run checks against a rebased version of the PR branch (assuming no merge conflicts exist)? I don't seem to recall that happening before.

In any case, I'll open another PR to fix the issue there and rebase this when that's in, but I'd be really curious to know what's happening here.

CAM-Gerlach commented:
I've made a tweak to the multiline checks (`Post-History` and `Authors`): they no longer require consistent indentation on continuation lines, only that at least one space be present after the newline (per RFC 2822). This makes naive (regex-based) parsing a tad more complicated, slightly increases the possibility of edge cases and doesn't look as clean in the source, but avoids the greater risk of being too nitpicky on authors for something that doesn't really matter that much.

hugovk commented on pep-0601.txt:
```diff
@@ -1,7 +1,7 @@
 Title: Forbid return/break/continue breaking out of finally
 Author: Damien George, Batuhan Taskaya
-Sponsor: Nick Coghlan
+Sponsor: Nick Coghlan <ncoghlan@gmail.com>
```
I asked Nick and he said it's fine too but the only reason he didn't add it is because he was just the sponsor here.

And checking [PEP 1](https://peps.python.org/pep-0001/#pep-header-preamble), only the author header mentions email. Sponsor and PEP-Delegate are listed as name only:

```
  Author: <list of authors' real names and optionally, email addrs>
* Sponsor: <real name of sponsor>
* PEP-Delegate: <PEP delegate's real name>
```

However, there are some existing PEPs with emails in Sponsor and *-Delegate...

- Sponsor with email: 31
- Sponsor without email: 12
- Delegate with email: 65
- Delegate without email: 41

Should PEP 1 be updated to allow optional emails in these headers? Or the emails removed? Something else?

CAM-Gerlach replied:
This is one of the (multiple) inconsistencies between PEP 1 and PEP 12 (and the template file), particularly when it comes to headers. [PEP 12 says](https://peps.python.org/pep-0012/#how-to-use-this-template)

```email
Author: [Full Name <email at example.com>]
Sponsor: *[Full Name <email at example.com>]
PEP-Delegate:
```

Since PEP 12 is more up to date and more precisely specified, I generally give it precedence, and I don't think there's a harm in allowing but not requiring an email to be specified, which is reflected in the updated checks in this PR.

> Should PEP 1 be updated to allow optional emails in these headers? Or the emails removed? Something else?

What I was going to do, as part of the larger updates planned, was to consolidate the repetitive and conflicting listings and descriptions of the headers and sections into PEP 12 and sync the template with it, focusing that PEP more tightly on the technical implementation details and author-focused guidance, while PEP 1 remains dedicated to describing the PEP _process_ and governance as a whole.

> Sponsor without email: 12

Huh; as a test I initially ran the check with the email required (which is how I found these), and it only identified four sponsors without emails, all of whom had them in multiple other PEPs (so I opportunistically updated them) as well as a few other issues. So it seems either your regex has false positives, or more concerning my regex (the one used here, just without the email optional) had false negatives, which is worrisome...

hugovk replied:
Updating PEPs 1 and 12: sounds reasonable.

Sponsor without email count: Yeah, quite likely some false positives, just did some quick greps for a rough comparison, so I wouldn't worry about it too much. Main point is there's quite a few _with_ email.

CAM-Gerlach replied:
To note, I initially planned to de-conflict them with this PR, or in a predecessor one, but decided it would be best to separate that off and probably do it as part of the planned consolidation, rather than duplicating effort three places.

hugovk commented on pep-0648.rst:
```diff
@@ -1,8 +1,7 @@
 Author: Mario Corchero <mariocj89@gmail.com>
-Sponsor: Pablo Galindo
-BDFL-Delegate: XXXX
+Sponsor: Pablo Galindo <pablogsal@python.org>
```
I checked with Pablo and he said let's add it here. 👍

pablogsal replied:
Thanks ❤️

CAM-Gerlach commented:
Updated the branch and fixed one issue (not allowing fragments in URLs).

Are there more things I need to change, or should we merge this? #2467 is otherwise ready to go and waiting on this one.

JelleZijlstra commented:
I approved, so I'm fine with merging.

hugovk commented:
Will give this another pass today!

hugovk reviewed (commented):
I still have some concerns about adding email addresses without permission, and replacing `user at example.com` with `user@example.com`, but everything else is fine, and I won't block merging over it. Thanks!

CAM-Gerlach commented:
> I still have some concerns about adding email addresses without permission,

Well again, the only email addresses added in the current version of the PR were those for same core devs who already chose to include them the headers of their other PEPs, and which are already listed in the [Authors Index](https://peps.python.org/pep-0000/#authors-owners), so it is only making things consistent between PEPs. Also, I'd initially thought from PEP 12 that the email address was a required element and had the checks enforce that for new PEPs, which is why I'd initially made the change, but I'd then realized after cross-referencing PEP 1 that it wasn't, and updated the checks accordingly.

Much the same is true of `user at example.com` -> `user@example.com`—its already obfuscated automatically, having it in the source makes authors (like myself at time) think they had to do it manually, and it modestly complicates the processing and checks, but likewise I quickly realized a relatively small simplification of the code was not worth the potential for more trouble for authors, so I ensured the checks and processing code allow it.

In both cases, there is little benefit conforming the existing source headers when neither is actually required; its just that likewise, it would also be a fair amount of work for no clear benefit to reverting them either, since the emails are already listed and obfuscated. But as it seems its the one thing that's blocking this PR, I guess I'll just bite the bullet and spend the time doing it.

CAM-Gerlach commented:
All done—this PR now does not add any existing emails, nor does it conform them to the standard address format. Also fixed what would be a conflict with #2531 , clarified the check feedback a bit further, and a couple other minor refinements.

hugovk reviewed (approved):
I appreciate all your work here, thank you!

CAM-Gerlach commented:
Actually, I noticed there's actually a significant problem as it stands allowing manual obfuscation—glancing at the `_mask_email()` source code as part of the trailing comma oversight you noted on #2531 and I then fixed in #2467 , I notice that the automatic obfuscation is actually a fair bit more effective than just replacing `@` with a literal ` at `; rather, it actually uses the raw HTML entity codes for some of the characters, significantly raising the bar for scrapers.

However, since emails with manual obfuscation applied don't get converted into reference nodes, the automatic obfuscation logic doesn't work, so these emails actually are substantially _less_ obfuscated than they would otherwise be (which PEP authors are almost certainly unaware of, and wasn't even to me until I dug deep into the code and actually tested it). Picking on myself here, from PEP 0, compare these two:

```html
<tr class="[row-odd]()"><td>Gerlach, C.A.M.</td>
<td>cam.gerlach at gerlach.cam</td>
</tr>
```

versus, e.g. 

```html
<tr class="[row-even]()"><td>van Rossum, Guido (GvR)</td>
<td>guido&#32;&#97;t&#32;python.org</td>
</tr>
```

I attempted to add support for also properly masking manually obfuscated emails, but after some testing realized it would require some pretty significant code changes due to how things are currently structured, as well as some hacky and potentially unreliable heuristics. Furthermore, I realized this also makes the not showing the email addresses all, or only as abbreviations, as discussed in #2514 more complicated, since whether or not the email is actually processed as an email changes the doctree structure and node types.

Therefore, I conclude it would be best to re-revert the part of the previous change to conform the small minority of emails that were manually obfuscated to use standard email syntax and have the inters flag deviations, so they are correctly processed and masked by the header transform code (and anything else that needs to mask/obfuscate/elide them), in order to ensure that the various automatic measures to protect authors' emails work consistently.

CAM-Gerlach commented:
Actually, upon giving this more thought, making the author-emails `abbr`s instead of literal text as discussed in #2514 still requires doing fairly involved string-munging anyway due to having to parse and transform  the older `Email (author)` format, so while it does make things a little more complicated, its not _that_ much worse than parsing a while different syntax.

As such, I suggest we just go ahead and merge this PR as-is with the manual obfuscation still untouched, and then I can address properly masking manually "obfuscated" emails along with formatting them consistently as `abbr`s in a followup PR. It would also be pretty easy to improve the obfuscation further by choosing different Unicode lookalike characters for the spaces and letters, which coupled with being embedded in the abbr and using raw character codes should make it virtually impossible for spam harvesting, far more so than the common and well known replacement of `@` with ` at `.

CAM-Gerlach commented:
Since it seems I've satisfied the immediate concerns that were raised and the two PEP editors that previously reviewed have :heavy_check_mark: ed, this PEP has been open for a while and it seems there aren't any further objections, and it is blocking some further discussed and agreed changes (`abbr` for emails, making Content-Type optional, etc.,), I'll go ahead and finally merge this now.
